keras_nlp/metrics/bleu.py

```diff
@@ -0,0 +1,449 @@
+    text. We find the "clipped count" of matching n-grams so as to not
+    give a high score to a (reference, prediction) pair with redundant, repeated
+    tokens. Secondly, BLEU score tends to reward shorter predictions more, which
+    is why a brevity penalty is applied to penalise short predictions.
+
+    Note on input shapes:
+    For `y_true` and `y_pred`, this class supports scalar values and batch
+    inputs of shapes `()`, `(batch_size,)` and `(batch_size, 1)`.
+
+    Args:
+        tokenizer: callable. A function that takes a string `tf.RaggedTensor`
+            (of any shape), and tokenizes the strings in the tensor. This
+            function should use TensorFlow graph ops. If the tokenizer is not
+            specified, the default tokenizer is used. The default tokenizer
+            replicates the behaviour of SacreBLEU's `"tokenizer_13a"` tokenizer
+            (https://github.com/mjpost/sacrebleu/blob/v2.1.0/sacrebleu/tokenizers/tokenizer_13a.py).
+        max_order: int. The maximum n-gram order to use. For example, if
+            `max_order` is set to 3, unigrams, bigrams, and trigrams will be
+            considered. Defaults to 4.
+        smooth: bool. Whether to apply Lin et al. 2004 smoothing to the BLEU
+            score. Defaults to False.
+        variant: string. Either `"corpus_bleu"` or `"sentence_bleu"`. The former
+            computes micro-average precision, which is equivalent to passing all
+            samples (across batches) all at once. In other words, summing the
+            numerators and denominators for each hypothesis-reference(s) pairs
+            before the division (in order to calculate precision). The latter is
+            the macro-averaged BLEU score which means that it computes the  BLEU
+            score for every sample separately and averages over these scores.
+            Defaults to `"corpus_bleu"`.
+        dtype: string or tf.dtypes.Dtype. Precision of metric computation. If
+               not specified, it defaults to tf.float32.
+        name: string. Name of the metric instance.
+        **kwargs: Other keyword arguments.
+
+    References:
+        - [Papineni et al., 2002](https://aclanthology.org/P02-1040/)
+    """
+
+    def __init__(
+        self,
+        tokenizer=None,
+        max_order=4,
+        smooth=False,
+        variant="corpus_bleu",
+        dtype=None,
+        name="bleu",
+        **kwargs,
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+
+        if not tf.as_dtype(self.dtype).is_floating:
+            raise ValueError(
+                "`dtype` must be a floating point type. "
+                f"Received: dtype={dtype}"
+            )
+
+        if variant not in ("corpus_bleu", "sentence_bleu"):
+            raise ValueError(
+                "`variant` must be either 'corpus_bleu' or 'sentence_bleu'. "
+                f"Received: variant={variant}"
+            )
+
+        def default_tokenizer(inputs):
```

Comment:
There are a lot of nested function that are quite long all over. Not very readable.

Can we either pull them out of the class entirely, or make them class methods?

Reply:
Made a few of them class methods.

Reply:
Maybe make this one a private method called `_tokenizer` that does.

```
if self.tokenizer:
    return self.tokenizer(x)

for pattern, replacement...
```

That way saving self.tokenizer in the config would work as expected.